### PR TITLE
[Fix #3138] Fix encoding mismatch issue

### LINKED ISSUE #3138: invalid byte sequence in US-ASCII .. reproducable in docker container
Hello.

Below is a git repo containig a simple project and a readme describing how to create a docker container and then run rubocop in it. When one does one can reproduce the following error message:

"invalid byte sequence in US-ASCII"

Manual investigation shows there are some non-ascii characters in the .rubocop.yml that are causing trouble (it would have been very helpful if rubocop had at least said what file was causing the trouble, or even better which line)

If I run rubocop on my development machine with the same file (ie not within docker) then I dont get this error.

https://github.com/dmoran-epages/rubocop-docker-error

I see that this error message has popped up quite a few times, but doesn't seem to be fully resolved. Maybe this same can help.

Cheers, Declan

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Fixes #3138. Catching the bug in a test is a bit cumbersome but I think it works.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it)
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -15,8 +15,14 @@ require 'rake'
 require 'rspec/core'
 require 'rspec/core/rake_task'
 require 'rubocop/rake_task'
+
 RSpec::Core::RakeTask.new(:spec)
 
+RSpec::Core::RakeTask.new(:ascii_spec) do |t|
+  t.ruby_opts = '-E ISO-8859-1'
```
Comment (owst) [inline Rakefile]: The rake task name doesn't line up with the encoding - could the encoding be `ASCII` rather than Latin1?

Comment (deivid-rodriguez) [inline Rakefile]: Absolutely, updated and rebased.

File: spec/rubocop/config_loader_spec.rb [hunk 2]
```diff
@@ -386,6 +387,14 @@
       )
     end
 
+    it 'loads configuration properly when it includes non-ascii characters ' do
+      create_file(configuration_path, ['# All these cops of mine are ❤',
```
Comment (mikegee) [inline spec/rubocop/config_loader_spec.rb]: Can you duplicate the issue by appending `.force_encoding('binary')` on this one literal string?

Change the default encoding of all the spec files seems a bit heavy-handed.

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: I don't think so, but you can try! :)

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: @mikegee I don't know what would be the rationale behind that `.force_encoding('binary')`, but I just tried it and it doesn't catch the bug.

The problem here is with the external encoding, and the only way to set that is through environment variables (`LANG`, `LC_ALL`) before running ruby, or the ruby flag `-E`. This option looked heavy-handed to me at first, but after looking at the final patch, I no longer think so.

I also tried setting [Enconding.default_external](http://ruby-doc.org/core-2.3.1/Encoding.html#method-c-default_external-3D) manually, but not only it's explicitly discouraged in the docs, but it also didn't catch the bug.

Comment (mikegee) [inline spec/rubocop/config_loader_spec.rb]: I was hoping for a quick win without understanding this issue completely.

This is tricky problem. If I just take the test case by itself, it passes as written. If I change Ruby's default external encoding too, like you did in the rake config, I get an error in the spec itself:
```
     Encoding::UndefinedConversionError:
       U+2764 from UTF-8 to US-ASCII
     # ./spec/support/file_helper.rb:20:in `write'
     # ./spec/support/file_helper.rb:20:in `puts'
     # ./spec/support/file_helper.rb:20:in `block in create_file'
     # ./spec/support/file_helper.rb:13:in `open'
     # ./spec/support/file_helper.rb:13:in `create_file'
     # ./spec/rubocop/config_loader_spec.rb:391:in `block (3 levels) in <main>'
```

You don't get that?

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: No, never seen that. How can I reproduce it? What's the command you're running?

Comment (mikegee) [inline spec/rubocop/config_loader_spec.rb]: Sorry, i realized I should have included the command in my previous comment after posting it. Its there now.

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: No problem. No, it passes for me. I googled the error and  `1.9.3` and OSX appeared. Running any of those?

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: What's the behavior for you _without_ the patch? Do you get the same error as now, or the error in #3138?

Comment (mikegee) [inline spec/rubocop/config_loader_spec.rb]: I'm running Ruby 2.0 on OSX.

To be clear my testing was your new spec example by itself, run `rspec` as normal: it passes. Then run the test with ascii as the default external encoding with that command I posted: error from spec. In neither case do I get an error from `ConfigLoader`'s implementation.

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: Not sure if I completely understand the results you're getting. Could you post the ouput of `bundle exec ascii_spec` and `bundle exec ruby -Ispec -rrspec/autorun -E ascii ./spec/rubocop/config_loader_spec.rb`, both _with_ and _without_ the following patch?

```diff
diff --git a/lib/rubocop/config_loader.rb b/lib/rubocop/config_loader.rb
index ec2a11e..abc4248 100644
--- a/lib/rubocop/config_loader.rb
+++ b/lib/rubocop/config_loader.rb
@@ -139,7 +139,7 @@ module RuboCop
       end
 
       def load_yaml_configuration(absolute_path)
-        yaml_code = IO.read(absolute_path)
+        yaml_code = IO.read(absolute_path, encoding: 'UTF-8')
         # At one time, there was a problem with the psych YAML engine under
         # Ruby 1.9.3. YAML.load_file would crash when reading empty .yml files
         # or files that only contained comments and blank lines. This problem
```

Comment (mikegee) [inline spec/rubocop/config_loader_spec.rb]: Sorry, I goofed up. I had started a what I thought was a different solution (changing the external encoding in `spec/support/file_helper.rb`).

I get the error you reported `invalid byte sequence in US-ASCII` without this fix but not with it.

Comment (deivid-rodriguez) [inline spec/rubocop/config_loader_spec.rb]: That's great to hear! :)

**Comment by bbatsov** [comment]
Has to be rebased.

**Comment by deivid-rodriguez** [comment]
Done!

**Pull request merged**